python/bindings/exec.cpp

```diff
@@ -0,0 +1,84 @@
+#include <pybind11/functional.h>
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "reffine/utils/utils.h"
+
+using namespace std;
+using namespace reffine;
+
+namespace py = pybind11;
+
+string to_string(shared_ptr<Func> fn) { return IRPrinter::Build(fn); }
+
+template <typename Output, typename... Inputs>
+Output execute_query(void* query, Output output, Inputs... inputs)
+{
+    using FuncType = void (*)(Output, Inputs...);
+    auto query2 = reinterpret_cast<FuncType>(query);
+    query2(output, (inputs)...);
+
+    return output;
+}
+
+PYBIND11_MODULE(exec, m)
+{
+    m.def("to_string", [](std::shared_ptr<Func> fn) { return to_string(fn); });
+
+    m.def("execute_query",
+          [](void* fn, py::array output, std::vector<py::array> inputs) {
+              auto output_buf = output.request();
+              auto output_ptr = output_buf.ptr;
+
+              std::vector<void*> input_ptrs;
+              for (auto& input : inputs) {
+                  auto input_buf = input.request();
+                  input_ptrs.push_back(input_buf.ptr);
+              }
+
+              if (input_ptrs.size() == 1) {
+                  execute_query(fn, output_ptr, output_ptr, input_ptrs[0]);
+              } else if (input_ptrs.size() == 2) {
+                  execute_query(fn, output_ptr, output_ptr, input_ptrs[0],
+                                input_ptrs[1]);
+              } else if (input_ptrs.size() == 3) {
+                  execute_query(fn, output_ptr, output_ptr, input_ptrs[0],
+                                input_ptrs[1], input_ptrs[2]);
+              } else {
+                  throw std::runtime_error("Unsupported number of inputs.");
+              }
+
+              return output;
+          });
+
+    m.def("execute_query", [](void* fn, py::capsule output,
```

Comment:
why do we need 2 execute_query functions? Also what is the difference between py::capsule and py:array?